data models setup for individual market application

saikumar9 opened the pull request:
# [Individual Market Application Data Models Design Link](https://www.figma.com/board/7OiSg4RXLTp7rABiVoAGxv/QHP-Application-Redesign?node-id=17-716&t=U1HonMGWx1F0XJw5-0)

# PR Checklist

Please check if your PR fulfills the following requirements:
- [x] The title follows our [guidelines](https://github.com/ideacrew/enroll/blob/trunk/CONTRIBUTING.md#commit)
- [x] Tests for the changes have been added (for bug fixes/features), and they use `let` helpers and `before` blocks.
- [ ] For all UI changes, there is Cucumber coverage.
- [ ] Any endpoint touched in the PR has an appropriate Pundit policy. For open endpoints, the reasoning is documented in the PR and code.
- [ ] Any endpoint modified in the PR only responds to the expected MIME types.
- [ ] For all scripts or rake tasks, how to run them is documented in both the PR and the code.
- [ ] There are no inline styles added.
- [ ] There is no inline JavaScript added.
- [ ] There is no hard-coded text added/updated in helpers/views/JavaScript. New/updated translation strings do not include markup/styles unless there is supporting documentation.
- [ ] Code does not use `.html_safe`.
- [ ] All images added/updated have alt text.
- [x] Does not bypass RuboCop rules in any way.

# PR Type
What kind of change does this PR introduce?:

- [ ] Bugfix
- [x] Feature (requires Feature flag)
- [ ] Data fix, Migration, or Report (inert code, no impact until run)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Dependency updates (e.g., add a new gem or update to a version)

# What is the ticket # detailing the issue?

Ticket: [CU-868d5whhn](https://app.clickup.com/t/868d5whhn)

# A brief description of the changes:

Current behavior: `N/A`

New behavior: Initial Model setup for CR-120+

# Feature Flag

For all new feature development, a feature flag is required to control the exposure of the feature to our end users. A feature flag needs a corresponding environment variable to initialize the state of the flag. Please share the name of the environment variable below that would enable/disable the feature and indicate which client(s) it applies to.

Variable name:

- [ ] DC
- [ ] ME

# Additional Context
Include any additional context that may be relevant to the peer review process.

saikumar9 committed:
wip - initial commit for setting up models

saikumar9 committed:
brings latest evidence code

saikumar9 committed:
fixes a typo

saikumar9 committed:
moved pending items into its own file

saikumar9 committed:
fix the rubocop issue

saikumar9 committed:
adds factories and tests for application

saikumar9 committed:
adds more tests

saikumar9 committed:
adds more tests

saikumar9 committed:
fixes rubocop raised error

saikumar9 committed:
adds V3 namespace for the new eligibility evidence pattern

saikumar9 committed:
adds missing frozen string literal true

saikumar9 committed:
updates associations to match with the new V3 namespace

saikumar9 committed:
adds documentation for V3 module

saikumar9 committed:
more tests and model changes

saikumar9 committed:
adds more logic and associations

saikumar9 committed:
adds more evidences with documentation

saikumar9 committed:
removes undefined module inclusion

saikumar9 committed:
removes duplicate financial_assistance tests from rspec job

saikumar9 committed:
more data model changes and tests

saikumar9 committed:
adds tests for alive evidence model

saikumar9 committed:
adds more tests

saikumar9 committed:
adds more tests for evidences

saikumar9 committed:
more tests

saikumar9 committed:
modifies tests

saikumar9 committed:
adds more tests

saikumar9 committed:
adds exhibit and some tests

saikumar9 committed:
adds more tests

saikumar9 committed:
adds more tests

saikumar9 committed:
wip - models for contact detail

kristinmerbach commented on app/models/contact_center/contact_detail.rb:
```diff
@@ -0,0 +1,111 @@
+# frozen_string_literal: true
+
+module ContactCenter
```
Let's discuss the naming of this

kristinmerbach commented on app/models/eligibilities/v3/exhibit.rb:
```diff
@@ -0,0 +1,19 @@
+    class Exhibit
+      include Mongoid::Document
+      include Mongoid::Timestamps
+      include ::HasDocument
```
I don't think top level exhibit needs this has_document as the only type that involves documents is the upload_exhibit

kristinmerbach commented on app/models/family.rb:
```diff
@@ -58,6 +58,9 @@ class Family
+  has_many :applications, class_name: 'Sbm::Application'
+
   # Collection of insured:  employees, consumers, residents
 
```
add method to pull latest determined application of any type. also will likely need to pull the type of latest determined

kristinmerbach commented on app/models/individual_market/attestation.rb:
```diff
@@ -0,0 +1,31 @@
+    # field :report_change_terms, type: Boolean
+    # field :parent_living_out_of_home_terms, type: Boolean
+    # field :attestation_terms, type: Boolean
+    # field :submission_terms, type: Boolean
```
do we want to include something for signature? right now required in UI etc. this might not be useful now, but if we build in for helpers / assisters to sign separately might be good for tracking etc down the line

mdkaraman replied:
Possibly worth noting-- the ATP schema makes this distinction, where signature is modeled as it's own thing b/c it may be different from primary person/applicant (like you're pointing out due to helpers/assisters involvement).  Could always cross this bridge later if we need to communicate the signer info to another system down the line.

kristinmerbach commented on app/models/person_name.rb:
```diff
@@ -0,0 +1,67 @@
+  # Defines the polymorphic relationship with the parent document
+  # This allows PersonName to be embedded in any model that sets itself as :person_nameable
+  embedded_in :person_nameable, polymorphic: true
+
```
if we're redoing names, might want to have a quick talk about the names of these fields -> more proper is things like given name rather than first, family name rather than last.

kristinmerbach reviewed (commented):
just started leaving some random thoughts, but so far looks good

saikumar9 committed:
adds contact profile models and tests

saikumar9 committed:
renames first_name and last_name to given_name and family_name

saikumar9 committed:
adds methods to pull the latest determined application for a given family

saikumar9 committed:
fixes a couple of rubocop raised issues

saikumar9 committed:
adds a new TODO

saikumar9 committed:
adds tests and changes validation

saikumar9 committed:
adds unused grant model

saikumar9 committed:
adds cascade_callbacks to set timestamps on embedded documents

we must always save the top level document and not embedded documents to avoid duplicate callbacks

saikumar9 committed:
adds tests for request_results and verification_histories

saikumar9 committed:
adds a placeholder model for immigration information

saikumar9 committed:
remove unwanted documentation

saikumar9 committed:
renames immigration_document to immigration_information

saikumar9 committed:
refactors some methods for preferences

kristinmerbach commented on components/financial_assistance/app/models/financial_assistance/applicant.rb:
```diff
@@ -1545,8 +1553,27 @@ def has_citizen_immigration_status?
+      @aptc_csr_eligibility = eligibilities.where(_type: 'Eligibilities::V3::AptcCsrEligibility').first
+    end
+
+    def individual_market_eligibility
```
individual market may be renamed to exchangeeligibility to match cms

kristinmerbach commented on components/financial_assistance/app/models/financial_assistance/application.rb:
```diff
@@ -242,6 +242,19 @@ class Application # rubocop:disable Metrics/ClassLength TODO: Remove this
+    scope :newest_determined_by_family_id, lambda { |family_id|
+      where(aasm_state: 'determined', family_id: family_id).order(created_at: :desc).limit(1)
+    }
+
```
on diagram we have has many tax households on this model - not here

kristinmerbach commented on app/models/eligibilities/v3/determinations/individual_market_determination.rb:
```diff
@@ -0,0 +1,20 @@
+# frozen_string_literal: true
```
CMS calls this the "ExchangeEligibility" worth discussing whether we match them or go with individual

saikumar9 committed:
ues basis as model name instead of criterion to match with CMS

saikumar9 committed:
adds more todos

saikumar9 committed:
Merge remote-tracking branch 'origin/trunk' into ivl_application_data_models

saikumar9 committed:
Merge remote-tracking branch 'origin/trunk' into ivl_application_data_models

saikumar9 committed:
Merge branch 'trunk' into ivl_application_data_models

saikumar9 committed:
Merge branch 'trunk' into ivl_application_data_models

mdkaraman commented on app/models/eligibilities/v3/verification_history.rb:
```diff
@@ -0,0 +1,32 @@
+
+module Eligibilities
+  module V3
+    # Stores the history of each transaction a user or admin has performed on an evidence
```
Technically might be more accurate to say:
`Stores the history of each transaction a user, admin, or automated process has performed on an evidence`
Given that PDM processes update these records as well.

mdkaraman commented on app/models/family.rb:
```diff
@@ -1683,8 +1686,46 @@ def none_applying_coverage?
     active_family_members.none?(&:is_applying_coverage)
   end
 
+  # Returns the most recently created determined application, whether it's an SBM or FAA application.
```
This comment explains the behavior of the method, but not why we'd want it to behave that way.  Why would we distinguish between the two determined apps using the created_on field? Would the sbm_app always have an older created_at timestamp because it is created first in the user flow?

mdkaraman replied:
So Financial Assistance Application contains all the info that Individual Market Application PLUS the info relevant to financial assistance.  If someone is new to the system and they opt to apply for assistance, they will begin with a Financial Assistance Application and not even have a Individual Market Application. (just noting the difference with how this works compared to current FAA implementation)

mdkaraman commented on app/models/individual_market/attestation.rb:
```diff
@@ -0,0 +1,31 @@
+# frozen_string_literal: true
+
+module IndividualMarket
+  # Represents a collection of attestations made by an applicant during the enrollment process
```
Can an admin attest on behalf of a user?  Do we need to track that distinction somewhere?  The wording in here implies that it's always the account holder who is the user attesting to something.

mdkaraman commented on app/models/individual_market/immigration_information.rb:
```diff
@@ -0,0 +1,67 @@
+    #   @return [IndividualMarket::Applicant] The applicant to which this immigration information belongs to
+    embedded_in :applicant, class_name: 'IndividualMarket::Applicant'
+
+    # Enable the commented fields below or add new fields (more descriptive) as needed.
```
Glad to see this commented out, as it probably makes sense to reconsider the way we have this implemented in the legacy models.  Dumping every field into a single model makes distinguishing which type of document has which relevant fields a real pain.  We could use a more thoughtful approach here.

mdkaraman commented on components/financial_assistance/app/models/financial_assistance/application.rb:
```diff
@@ -242,6 +243,19 @@ class Application # rubocop:disable Metrics/ClassLength TODO: Remove this
     index({ "applicants.local_mec_evidence.aasm_state" => 1 })
 
+    # @!index [Hash] Creates a compound index on aasm_state, family_id, and created_at fields
+    # @param aasm_state [Integer] The application state, with 1 indicating ascending order
```
Possibly overkill having comments on the indexes?

mdkaraman reviewed (commented):
Clear that a huge amount of thought and work went into this, incredible job!  I think the class-level comments are particular strong and helpful, along with the use of memoization and thoughtful relationships/data structures throughout the new models.  I think some of the method-level comments may be a bit verbose without adding much value.  I left some other thoughts sprinkled throughout.  Nothing flagged as critical, feel free to respond and resolve as you see fit.

saikumar9 committed:
Merge branch 'trunk' into ivl_application_data_models

saikumar9 committed:
Merge branch 'trunk' into ivl_application_data_models

mdkaraman commented on app/models/contact_profile/contact_detail.rb:
```diff
@@ -0,0 +1,98 @@
+
+    # Retrieves the most recently added email
+    # @return [ContactProfile::Email, nil] The newest email or nil if none exists
+    def email
```
Ok that makes sense. So practically speaking, we'd be using the other memoized methods like `personal_email` to access the specific email we're interested in.  This `email` method could come in handy for inspecting the wrapper object or finding old versions if we need to.

mdkaraman commented on app/models/contact_profile/phone.rb:
```diff
@@ -0,0 +1,29 @@
+
+    # @!attribute [rw] mobile_phone
+    #   @return [ContactProfile::MobilePhone] The embedded mobile phone details
+    embeds_one :mobile_phone, class_name: 'ContactProfile::MobilePhone'
```
Yeah, I can't see a way to avoid having separate models now that I understand this better.  I guess we could abstract the common fields out into a module or something, but that is an implementation deal not worth quibbling over.

saikumar9 committed:
Merge branch 'trunk' into ivl_application_data_models

saikumar9 committed:
Merge branch 'trunk' into ivl_application_data_models

saikumar9 committed:
Merge branch 'trunk' into ivl_application_data_models

saikumar9 committed:
Merge branch 'trunk' into ivl_application_data_models

kristinmerbach reviewed (approved):
LGTM!

saikumar9 committed:
Merge branch 'trunk' into ivl_application_data_models

mdkaraman reviewed (approved):
Looks good!  Nice work 🔥
